src/networking.c

```diff
@@ -1347,6 +1352,105 @@ void unlinkClient(client *c) {
     if (c->flags & CLIENT_TRACKING) disableTracking(c);
 }
 
+/* Clear the client state to resemble a newly connected client. */
+void clearClientConnectionState(client *c) {
+    listNode *ln;
+
+    /* MONITOR clients are also marked with CLIENT_SLAVE, we need to
+     * distinguish between the two.
+     */
+    if (c->flags & CLIENT_MONITOR) {
+        ln = listSearchKey(server.monitors,c);
+        serverAssert(ln != NULL);
+        listDelNode(server.monitors,ln);
+
+        c->flags &= ~(CLIENT_MONITOR|CLIENT_SLAVE);
+    }
+
+    serverAssert(!(c->flags &(CLIENT_SLAVE|CLIENT_MASTER)));
+
+    if (c->flags & CLIENT_TRACKING) disableTracking(c);
+    selectDb(c,0);
+    c->resp = 2;
+
+    clientSetDefaultAuth(c);
+    moduleNotifyUserChanged(c);
+    discardTransaction(c);
+
+    pubsubUnsubscribeAllChannels(c,0);
+    pubsubUnsubscribeAllPatterns(c,0);
+
+    if (c->name) {
+        decrRefCount(c->name);
+        c->name = NULL;
+    }
+
+    /* Selectively clear state flags not covered above */
+    c->flags &= ~(CLIENT_ASKING|CLIENT_READONLY|CLIENT_PUBSUB|
+                  CLIENT_REPLY_OFF|CLIENT_REPLY_SKIP_NEXT);
+}
+
+/* Clear the client state fully and initialize it as a new client object. This
+ * function is useful when you want to reuse the client object to avoid
+ * costly freeClient()/createClient() calls. */
+void clearClient(client *c) {
+    serverAssert(!(c->flags & CLIENT_PROTECTED));
+
+    /* For connected clients, call the disconnection event of modules hooks. */
+    if (c->conn) {
+        moduleFireServerEvent(REDISMODULE_EVENT_CLIENT_CHANGE,
+                              REDISMODULE_SUBEVENT_CLIENT_CHANGE_DISCONNECTED,
+                              c);
+    }
+
+    clearClientConnectionState(c);
+
+    /* Deallocate structures used to block on blocking ops. */
+    if (c->flags & CLIENT_BLOCKED) unblockClient(c);
+    dictEmpty(c->bpop.keys, NULL);
+
+    sdsclear(c->querybuf);
+    /* Resize querybuf if it is too big. If this client ends up in a cache
+     * after being cleared, it may go undetected with a large query buffer,
+     * possibly wasting memory. */
+    if (sdsavail(c->querybuf) > PROTO_RESIZE_THRESHOLD) {
+        c->querybuf = sdsResize(c->querybuf, PROTO_RESIZE_THRESHOLD);
+    }
+
+    sdsclear(c->pending_querybuf);
+    listEmpty(c->watched_keys);
+    dictEmpty(c->pubsub_channels, NULL);
+    listEmpty(c->pubsub_patterns);
+    listEmpty(c->reply);
+    freeReplicaReferencedReplBuffer(c);
+    freeClientArgv(c);
+    freeClientOriginalArgv(c);
+
+    /* Unlink the client: this will close the socket, remove the I/O
+     * handlers, and remove references of the client from different
+     * places where active clients may be referenced. */
+    unlinkClient(c);
```

Comment:
same here. not sure if this belongs here or not.
we need to define the role of this function better.
i.e. we never add clients with no `conn` to the client list, so no need to remove them from there, or unregister their event handlers.

Reply:
I just tried to create a generic function to clear clients. Clients can be with or without `conn`. So, I thought we want a single function to `clear()` client objects which can be real clients or fake clients.

Reply:
but what would we use such a function for?
resetCommand can't obviously use it (doesn't want to disconnect the client or remove it from the client list).
maybe if we wanted to recycle the client structs for normal clients (i.e. when they disconnect add them to a pool).
i'm actually not sure it'll be a big performance win, not sure why client struct creation takes so long, maybe it's registration part (in the different lists) is even heavier.

Reply:
Yes, there is no use case for real clients for now. You asked for a single function to clear clients previously in the comments, so I created one `clear` function that can clear all kind of clients. Looks like I misunderstood, you were talking about fake clients only? 

client create/free involves 8 malloc/free pairs including client's own malloc. That takes quite time. So, I tried to avoid freeing memory in clearClient() as much as possible.

Reply:
i think my comment about a single clear function was to share code with resetCommand.